Fix more NumberToStringBuffer conversions to use std::span

ddkilzer opened the pull request:
#### 058660103448a14477572e3509a632991934c777
<pre>
Fix more NumberToStringBuffer conversions to use std::span
&lt;<a href="https://bugs.webkit.org/show_bug.cgi?id=280596">https://bugs.webkit.org/show_bug.cgi?id=280596</a>&gt;
&lt;<a href="https://rdar.apple.com/136944262">rdar://136944262</a>&gt;

Reviewed by Dan Glastonbury, Darin Adler, and Mike Wyrzykowski.

Updating WTF::numberTo...String...() functions in dtoa.{cpp,h} to return
std::span values frequently eliminates a call to strlen() to get the
length of the C-string bytes stored in the std::array.

* Source/JavaScriptCore/runtime/NumberPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION(numberProtoFuncToExponential, ...)):
- Stop calling String::fromLatin1() since StringBuilder::Finalize()
  returns a std::span now.  Eliminates a call to strlen().
* Source/JavaScriptCore/runtime/PropertyName.h:
(JSC::isCanonicalNumericIndexString):
- Simplify now that WTF::numberToStringAndSize() returns std::span.

* Source/WTF/wtf/dragonbox/dragonbox_to_chars.h:
(WTF::dragonbox::ToExponential):
(WTF::dragonbox::ToShortest):
- Change ASSERT() to static_assert() when possible to catch bugs
  earlier.
* Source/WTF/wtf/dtoa.cpp:
(WTF::numberToStringAndSize):
(WTF::numberToStringWithTrailingPoint):
(WTF::numberToFixedPrecisionString):
(WTF::numberToFixedWidthString):
(WTF::numberToCSSString):
- Update to return std::span.
* Source/WTF/wtf/dtoa.h:
(WTF::numberToStringAndSize):
(WTF::numberToStringWithTrailingPoint):
(WTF::numberToFixedPrecisionString):
(WTF::numberToFixedWidthString):
(WTF::numberToCSSString):
- Add LIFETIME_BOUND attribute since std::span return value depends on
  the lifetime of the NumberToStringBuffer&amp; argument.
- Update to return std::span.
* Source/WTF/wtf/dtoa/double-conversion.cc:
(WTF::double_conversion::DoubleToStringConverter::CreateExponentialRepresentation):
- Make sure length is always valid.
- Change while() loop to check value of first_char_pos as well.
(WTF::double_conversion::DoubleToStringConverter::CreateDecimalRepresentation):
(WTF::double_conversion::DoubleToStringConverter::ToShortestIeeeNumber):
(WTF::double_conversion::DoubleToStringConverter::ToFixed):
(WTF::double_conversion::DoubleToStringConverter::ToFixedUncapped):
(WTF::double_conversion::DoubleToStringConverter::ToExponential):
(WTF::double_conversion::DoubleToStringConverter::ToPrecision):
(WTF::double_conversion::StringToIeee):
- Change `const int` buffer lengths to `constexpr size_t`.
- Change ASSERT() to static_assert() when possible to catch bugs
  earlier.
- Change ASSERT() to ASSERT_WITH_SECURITY_IMPLICATION() to catch bugs.
* Source/WTF/wtf/dtoa/utils.h:
(WTF::double_conversion::StrLength):
(WTF::double_conversion::BufferReference::BufferReference):
(WTF::double_conversion::BufferReference::SubBufferReference):
(WTF::double_conversion::BufferReference::operator[] const):
(WTF::double_conversion::StringBuilder::position const):
(WTF::double_conversion::StringBuilder::AddCharacter):
(WTF::double_conversion::StringBuilder::AddSubstring):
(WTF::double_conversion::StringBuilder::RemoveCharacters):
- Change ASSERT() to ASSERT_WITH_SECURITY_IMPLICATION() to catch bugs.
(WTF::double_conversion::StringBuilder::Finalize):
- Never write out-of-bounds even if position_ is negative.
- Change to return std::span instead of pointer.
* Source/WTF/wtf/text/AtomString.cpp:
(WTF::AtomString::number):
- Simplify now that WTF::numberToStringAndSize() returns std::span.
* Source/WTF/wtf/text/StringConcatenateNumbers.h:
(WTF::StringTypeAdapter::StringTypeAdapter):
(WTF::FormattedNumber::fixedPrecision):
(WTF::FormattedNumber::fixedWidth):
(WTF::FormattedCSSNumber::create):
- Set m_length from std::span::size().
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::number):
* Source/WebCore/platform/Decimal.cpp:
(WebCore::Decimal::fromDouble):
- Simplify now that WTF::numberToStringAndSize() returns std::span.

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
(WGSL::Metal::FunctionDefinitionWriter::serializeConstant):
- Simplify now that WTF::numberToStringAndSize() returns std::span.
- Avoiding use of WTF::span(const char*) eliminates a call to strlen().

* Tools/TestWebKitAPI/Tests/WTF/DragonBoxTest.cpp:
(TestWebKitAPI::DragonBoxTest::dragonBoxToString):
(TestWebKitAPI::DragonBoxTest::dragonBoxToExponential):
(TestWebKitAPI::DragonBoxTest::doubleConversionToString):
(TestWebKitAPI::DragonBoxTest::doubleConversionToExponential):
- Update for std::span returned from StringBuilder::Finalize() method.
- Assert that length returned from Finalize() matches strlen() of the
  buffer backed by the std::array.

Canonical link: <a href="https://commits.webkit.org/285520@main">https://commits.webkit.org/285520@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/765157dc7be86aebee12366768e89ceb242ef0a4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72853 "17 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52278 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25653 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77050 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24087 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74968 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60083 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23903 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57287 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15773 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75920 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47263 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62704 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37711 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43910 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20170 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22416 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/65986 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65765 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20527 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78722 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/72109 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/17099 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19671 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65731 "Found 32 new test failures: compositing/animation/repaint-after-clearing-shared-backing.html compositing/shared-backing/overlap-after-end-sharing.html fast/text/crash-letter-spacing-infinite.html fast/text/crash-word-spacing-infinite.html imported/w3c/web-platform-tests/css/css-backgrounds/animations/background-color-animation-half-opaque.html imported/w3c/web-platform-tests/css/css-backgrounds/animations/background-color-transparent-animation-in-body.html imported/w3c/web-platform-tests/css/css-view-transitions/new-content-ancestor-clipped.html imported/w3c/web-platform-tests/css/css-view-transitions/new-content-captures-clip-path.html imported/w3c/web-platform-tests/css/css-view-transitions/new-content-changes-overflow.html imported/w3c/web-platform-tests/css/css-view-transitions/new-content-container-writing-modes.html ... (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/17147 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62710 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65005 "Found 1 new API test failure: /WebKitGTK/TestUIClient:/webkit/WebKitWebView/open-window-default-size (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13318 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6972 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/93888 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11213 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/48076 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2863 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20659 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/49143 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50438 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48888 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -2224,23 +2224,23 @@ void FunctionDefinitionWriter::visit(AST::AbstractFloatLiteral& literal)
     WTF::numberToStringWithTrailingPoint(literal.value(), buffer);
 
-    m_stringBuilder.append(span(&buffer[0]));
+    m_stringBuilder.append(std::span { buffer });
```
```suggestion
    m_stringBuilder.append(std::span { buffer }.first(strlen(buffer.data())));
```

ddkilzer replied:
Moving the `std::span` construction into `WTF::numberToStringWithTrailingPoint()` was even better!

ddkilzer commented on Source/WTF/wtf/dtoa/utils.h:
```diff
@@ -304,12 +310,13 @@ class StringBuilder {
 
   // Finalize the string by 0-terminating it and returning the buffer.
-  char* Finalize() {
+  char* Finalize(size_t& length) {
```
This should just return a `std::span`.  Not sure why I didn't see it earlier!

ddkilzer commented on Source/WTF/wtf/dragonbox/dragonbox_to_chars.h:
```diff
@@ -137,22 +137,22 @@ typedef WTF::double_conversion::StringBuilder StringBuilder;
     char buffer[buffer_length];
     auto* cursor = detail::to_chars_n<Mode::ToExponential>(value, buffer);
-    result_builder->AddSubstring(buffer, cursor - buffer);
+    result_builder->AddSubstring({ buffer, std::min<size_t>(sizeof(buffer), static_cast<size_t>(cursor - buffer)) });
```
Today I learned that a `std::span` may be constructed directly from a ` char buffer[N]` (assuming `buffer` is not a variable-length array or VLA), so this becomes much nicer to read:

```diff
-     result_builder->AddSubstring({ buffer, std::min<size_t>(sizeof(buffer), static_cast<size_t>(cursor - buffer)) });
+     result_builder->AddSubstring(std::span { buffer }.first(cursor - buffer));
```

darinadler commented on Source/WTF/wtf/dragonbox/dragonbox_to_chars.h:
```diff
@@ -137,22 +137,22 @@ typedef WTF::double_conversion::StringBuilder StringBuilder;
 void ToExponential(Float value, StringBuilder* result_builder)
 {
-    ASSERT((std::is_same<Float, double>::value) || (std::is_same<Float, float>::value));
+    static_assert((std::is_same<Float, double>::value) || (std::is_same<Float, float>::value));
```
Seems like this expression has some extra parentheses, also we should use `is_same_v` to make it shorter.

darinadler commented on Source/WTF/wtf/dragonbox/dragonbox_to_chars.h:
```diff
@@ -137,22 +137,22 @@ typedef WTF::double_conversion::StringBuilder StringBuilder;
 void ToShortest(Float value, StringBuilder* result_builder)
 {
-    ASSERT((std::is_same<Float, double>::value) || (std::is_same<Float, float>::value));
+    static_assert((std::is_same<Float, double>::value) || (std::is_same<Float, float>::value));
```
Seems like this expression has some extra parentheses, also we should use `is_same_v` to make it shorter.

darinadler commented on Source/WTF/wtf/text/AtomString.cpp:
```diff
@@ -108,15 +108,15 @@ AtomString AtomString::number(unsigned long long number)
-    size_t length = numberToStringAndSize(number, buffer);
-    return AtomString { std::span { bitwise_cast<const LChar*>(buffer.data()), length } };
+    auto span = numberToStringAndSize(number, buffer);
+    return AtomString { spanReinterpretCast<const LChar>(span) };
```
I think we can use `byteCast<LChar>` in cases like this instead of `spanReinterpretCast<const LChar>`. I didn't point it out in the other cases but I think it's better in all these cases, unless I'm mistaken about it working.

ddkilzer replied:
Not sure how to use `byteCast<LChar>` here without re-introducing a `-Wunsafe-buffer-usage` warning.  (That's the reason `spanReinterpretCast<const LChar>()` is being used.)

darinadler replied:
Does not make logical sense to me, because `byteCast` should,be identical to `spanReinterpretCasr` in this respect. If it's not, I'd want to figure out why and fix it.

mwyrzykowski reviewed (approved):
Looks good to me outside the style warnings 👍

darinadler commented on Source/WTF/wtf/text/WTFString.cpp:
```diff
@@ -241,15 +241,15 @@ String String::numberToStringFixedPrecision(double number, unsigned precision, T
-    size_t length = numberToStringAndSize(number, buffer);
-    return String { std::span { buffer.data(), length } };
+    auto span = numberToStringAndSize(number, buffer);
+    return String { span };
```
Seems like this would read better as a one-liner without the local variable.

darinadler commented on Source/WTF/wtf/text/WTFString.cpp:
```diff
@@ -241,15 +241,15 @@ String String::numberToStringFixedPrecision(double number, unsigned precision, T
-    size_t length = numberToStringAndSize(number, buffer);
-    return String { std::span { buffer.data(), length } };
+    auto span = numberToStringAndSize(number, buffer);
+    return String { span };
```
Seems like this would read better as a one-liner without the local variable.

darinadler commented on Source/WebCore/platform/Decimal.cpp:
```diff
@@ -547,8 +547,8 @@ Decimal Decimal::fromDouble(double doubleValue)
-        size_t length = numberToStringAndSize(doubleValue, buffer);
-        return fromString(StringView { std::span { buffer.data(), length } });
+        auto span = numberToStringAndSize(doubleValue, buffer);
+        return fromString(StringView { span });
```
Seems like this would read better as a one-liner without the local variable.

darinadler commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -2222,25 +2222,22 @@ void FunctionDefinitionWriter::visit(AST::Unsigned32Literal& literal)
-
-    m_stringBuilder.append(span(&buffer[0]));
+    auto span = WTF::numberToStringWithTrailingPoint(literal.value(), buffer);
+    m_stringBuilder.append(span);
```
Seems like this would read better as a one-liner without the local variable.

darinadler commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -2222,25 +2222,22 @@ void FunctionDefinitionWriter::visit(AST::Unsigned32Literal& literal)
-
-    m_stringBuilder.append(span(&buffer[0]));
+    auto span = WTF::numberToStringWithTrailingPoint(literal.value(), buffer);
+    m_stringBuilder.append(span);
```
Seems like this would read better as a one-liner without the local variable.

darinadler commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -2222,25 +2222,22 @@ void FunctionDefinitionWriter::visit(AST::Unsigned32Literal& literal)
-
-    m_stringBuilder.append(span(&buffer[0]));
+    auto span = WTF::numberToStringWithTrailingPoint(literal.value(), buffer);
+    m_stringBuilder.append(span);
```
Seems like this would read better as a one-liner without the local variable.

darinadler commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -2559,20 +2556,20 @@ void FunctionDefinitionWriter::serializeConstant(const Type* type, ConstantValue
-                WTF::numberToStringWithTrailingPoint(std::get<double>(value), buffer);
-                m_stringBuilder.append(span(&buffer[0]));
+                auto span = WTF::numberToStringWithTrailingPoint(std::get<double>(value), buffer);
+                m_stringBuilder.append(span);
```
Seems like this would read better as a one-liner without the local variable.

darinadler commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -2559,20 +2556,20 @@ void FunctionDefinitionWriter::serializeConstant(const Type* type, ConstantValue
-                WTF::numberToStringWithTrailingPoint(std::get<float>(value), buffer);
-                m_stringBuilder.append(span(&buffer[0]));
+                auto span = WTF::numberToStringWithTrailingPoint(std::get<float>(value), buffer);
+                m_stringBuilder.append(span);
```
Seems like this would read better as a one-liner without the local variable.

darinadler commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -2559,20 +2556,20 @@ void FunctionDefinitionWriter::serializeConstant(const Type* type, ConstantValue
-                WTF::numberToStringWithTrailingPoint(std::get<half>(value), buffer);
-                m_stringBuilder.append(span(&buffer[0]), 'h');
+                auto span = WTF::numberToStringWithTrailingPoint(std::get<half>(value), buffer);
+                m_stringBuilder.append(span, 'h');
```
Seems like this would read better as a one-liner without the local variable.

Commit:
Fix more NumberToStringBuffer conversions to use std::span
<https://bugs.webkit.org/show_bug.cgi?id=280596>
<rdar://136944262>

Reviewed by Dan Glastonbury, Darin Adler, and Mike Wyrzykowski.

Updating WTF::numberTo...String...() functions in dtoa.{cpp,h} to return
std::span values frequently eliminates a call to strlen() to get the
length of the C-string bytes stored in the std::array.

* Source/JavaScriptCore/runtime/NumberPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION(numberProtoFuncToExponential, ...)):
- Stop calling String::fromLatin1() since StringBuilder::Finalize()
  returns a std::span now.  Eliminates a call to strlen().
* Source/JavaScriptCore/runtime/PropertyName.h:
(JSC::isCanonicalNumericIndexString):
- Simplify now that WTF::numberToStringAndSize() returns std::span.

* Source/WTF/wtf/dragonbox/dragonbox_to_chars.h:
(WTF::dragonbox::ToExponential):
(WTF::dragonbox::ToShortest):
- Change ASSERT() to static_assert() when possible to catch bugs
  earlier.
* Source/WTF/wtf/dtoa.cpp:
(WTF::numberToStringAndSize):
(WTF::numberToStringWithTrailingPoint):
(WTF::numberToFixedPrecisionString):
(WTF::numberToFixedWidthString):
(WTF::numberToCSSString):
- Update to return std::span.
* Source/WTF/wtf/dtoa.h:
(WTF::numberToStringAndSize):
(WTF::numberToStringWithTrailingPoint):
(WTF::numberToFixedPrecisionString):
(WTF::numberToFixedWidthString):
(WTF::numberToCSSString):
- Add LIFETIME_BOUND attribute since std::span return value depends on
  the lifetime of the NumberToStringBuffer& argument.
- Update to return std::span.
* Source/WTF/wtf/dtoa/double-conversion.cc:
(WTF::double_conversion::DoubleToStringConverter::CreateExponentialRepresentation):
- Make sure length is always valid.
- Change while() loop to check value of first_char_pos as well.
(WTF::double_conversion::DoubleToStringConverter::CreateDecimalRepresentation):
(WTF::double_conversion::DoubleToStringConverter::ToShortestIeeeNumber):
(WTF::double_conversion::DoubleToStringConverter::ToFixed):
(WTF::double_conversion::DoubleToStringConverter::ToFixedUncapped):
(WTF::double_conversion::DoubleToStringConverter::ToExponential):
(WTF::double_conversion::DoubleToStringConverter::ToPrecision):
(WTF::double_conversion::StringToIeee):
- Change `const int` buffer lengths to `constexpr size_t`.
- Change ASSERT() to static_assert() when possible to catch bugs
  earlier.
- Change ASSERT() to ASSERT_WITH_SECURITY_IMPLICATION() to catch bugs.
* Source/WTF/wtf/dtoa/utils.h:
(WTF::double_conversion::StrLength):
(WTF::double_conversion::BufferReference::BufferReference):
(WTF::double_conversion::BufferReference::SubBufferReference):
(WTF::double_conversion::BufferReference::operator[] const):
(WTF::double_conversion::StringBuilder::position const):
(WTF::double_conversion::StringBuilder::AddCharacter):
(WTF::double_conversion::StringBuilder::AddSubstring):
(WTF::double_conversion::StringBuilder::RemoveCharacters):
- Change ASSERT() to ASSERT_WITH_SECURITY_IMPLICATION() to catch bugs.
(WTF::double_conversion::StringBuilder::Finalize):
- Never write out-of-bounds even if position_ is negative.
- Change to return std::span instead of pointer.
* Source/WTF/wtf/text/AtomString.cpp:
(WTF::AtomString::number):
- Simplify now that WTF::numberToStringAndSize() returns std::span.
* Source/WTF/wtf/text/StringConcatenateNumbers.h:
(WTF::StringTypeAdapter::StringTypeAdapter):
(WTF::FormattedNumber::fixedPrecision):
(WTF::FormattedNumber::fixedWidth):
(WTF::FormattedCSSNumber::create):
- Set m_length from std::span::size().
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::number):
* Source/WebCore/platform/Decimal.cpp:
(WebCore::Decimal::fromDouble):
- Simplify now that WTF::numberToStringAndSize() returns std::span.

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
(WGSL::Metal::FunctionDefinitionWriter::serializeConstant):
- Simplify now that WTF::numberToStringAndSize() returns std::span.
- Avoiding use of WTF::span(const char*) eliminates a call to strlen().

* Tools/TestWebKitAPI/Tests/WTF/DragonBoxTest.cpp:
(TestWebKitAPI::DragonBoxTest::dragonBoxToString):
(TestWebKitAPI::DragonBoxTest::dragonBoxToExponential):
(TestWebKitAPI::DragonBoxTest::doubleConversionToString):
(TestWebKitAPI::DragonBoxTest::doubleConversionToExponential):
- Update for std::span returned from StringBuilder::Finalize() method.
- Assert that length returned from Finalize() matches strlen() of the
  buffer backed by the std::array.

Canonical link: https://commits.webkit.org/285520@main
